EIPS/eip-7709.md

```diff
@@ -0,0 +1,99 @@
+  if arg >= block.number or (arg + BLOCKHASH_SERVE_WINDOW) < block.number
+    return 0
+
+  # performs an sload on arg % HISTORY_SERVE_WINDOW including gas charges,
+  # warming effects as well as execution accesses
+  #
+  # note that the `BLOCKHASH_SERVE_WINDOW` and the 2935 ring buffer window
+  # `HISTORY_SERVE_WINDOW` for slot calculation are different
+  return state.load_slot(HISTORY_STORAGE_ADDRESS, arg % HISTORY_SERVE_WINDOW)
+```
+
+ONLY if the `arg` is within the correct `BLOCKHASH` window, clients can choose to either
+
+* do a direct `SLOAD` from state, or
+* do a system call to [EIP-2935](./eip-2935.md) contract `get` or
+* serve from memory or as per current designs if maintaining requisite history (full clients for e.g.)
+
+However the entire semantics and after effects of the `SLOAD` operation needs to be applied as per the current fork if the `arg` is within the correct `BLOCKHASH` window:
+
+* `SLOAD` gas costs (cold or warm) for the `arg % HISTORY_SERVE_WINDOW` slot.
+* `SLOAD` after effects on the slot (warming the slot)
+* `SLOAD` accesses added to execution witnesses if Verkle ([EIP-6800](./eip-6800.md) and [EIP-4762](./eip-4762.md)) is activated
+
+### Activation
+
+This EIP specifies the transition to the new logic assuming that [EIP-2935](./eip-2935.md) has been activated:
+
+* sufficiently ahead of this EIP's activation (>= `BLOCKHASH_SERVE_WINDOW`) or
+* at genesis for testnets/devnets where this EIP could also be activated at genesis
+
+The current proposal is to activate this EIP with Verkle to allow for stateless execution of the block.
+
+### Gas costs
+
+As described above, if the `arg` to be resolved is within the correct window, the corresponding `SLOAD` charges and accesses are to be applied for the slot `arg % HISTORY_SERVE_WINDOW`. Note that the `HISTORY_SERVE_WINDOW` and `BLOCKHASH_SERVE_WINDOW` are different.
+
+### System contract
+
+Even if the clients choose to resolve `BLOCKHASH` through system call to [EIP-2935](./eip-2935.md) contract, the gas cost for the system code execution (and also the code witnesses if Verkle activated) is not applied. Only the effect of `SLOAD` is applied as described above.
+
+## Rationale
+
+* The reason behind the updated gas cost is to match the real operation, which is equivalent to an `SLOAD`.
+* The [EIP-2935](./eip-2935.md) system contract execution charges (and accesses) are not applied to keep the gas low and to keep things simple for clients which choose to resolve `BLOCKHASH` in other ways (directly or though memory/maintained history)
+
+## Backwards Compatibility
+
+This EIP introduces a significant increase in the cost of `BLOCKHASH`, which could break use-cases that rely on the previous gas cost. Also, this EIP introduces a breaking change in the case where less than `BLOCKHASH_SERVE_WINDOW` elapse between the [EIP-2935](./eip-2935.md) fork and this EIP's fork as the [EIP-2935](./eip-2935.md) system contract would not have saved the required history.
```

Comment:
but the latter sentence is only relevant in the case that 2935 isn't activated at genesis.

You should also specify how it breaks, then: calling the contract would return 0, for any block prior to the 2935 block.

Reply:
we have already mentioned the activation dependency of 2935